ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/PullQueryExecutor.java

```diff
@@ -559,37 +563,43 @@ private static boolean isSelectStar(final Select select) {
       final KsqlExecutionContext executionContext,
       final ImmutableAnalysis analysis,
       final LogicalSchema outputSchema,
+      final Optional<WindowType> windowType,
       final QueryId queryId,
       final Stacker contextStacker
   ) {
-    final boolean referencesRowTime = analysis.getSelectColumnRefs().stream()
-        .anyMatch(ref -> ref.name().equals(SchemaUtil.ROWTIME_NAME));
-
-    final boolean referencesRowKey = analysis.getSelectColumnRefs().stream()
-        .anyMatch(ref -> ref.name().equals(SchemaUtil.ROWKEY_NAME));
+    final boolean noSystemColumns = analysis.getSelectColumnRefs().stream()
+        .noneMatch(ref -> SchemaUtil.systemColumnNames().contains(ref.name()));
 
     final LogicalSchema intermediateSchema;
-    final PreSelectTransformer preSelectTransform;
-    if (!referencesRowTime && !referencesRowKey) {
+    final Function<TableRow, GenericRow> preSelectTransform;
+    if (noSystemColumns) {
       intermediateSchema = input.schema;
-      preSelectTransform = (rowTime, key, value) -> value;
+      preSelectTransform = TableRow::value;
     } else {
       // SelectValueMapper requires the rowTime & key fields in the value schema :(
-      intermediateSchema = LogicalSchema.builder()
-          .keyColumns(input.schema.key())
-          .valueColumns(input.schema.value())
-          .valueColumns(input.schema.metadata())
-          .valueColumns(input.schema.key())
-          .build();
-
-      preSelectTransform = (rowTime, key, value) -> {
-        value.getColumns().add(rowTime);
-
-        key.schema().fields().forEach(f -> {
-          final Object keyField = key.get(f);
-          value.getColumns().add(keyField);
+      final boolean windowed = windowType.isPresent();
+
+      intermediateSchema = input.schema
+          .withMetaAndKeyColsInValue(windowed);
+
+      preSelectTransform = row -> {
+        final Struct key = row.key();
+        final List<Object> columns = row.value().getColumns();
+
+        columns.add(0, row.rowTime());
+
+        final List<Object> keyFields = key.schema().fields().stream()
+            .map(key::get)
+            .collect(Collectors.toList());
+
+        columns.addAll(1, keyFields);
+
+        row.window().ifPresent(window -> {
+          columns.add(1, window.start().toEpochMilli());
+          columns.add(2, window.end().toEpochMilli());
```

Comment:
just sanity checking, the reason we're adding this at 1 and 2 is so that they come after the ROWTIME but before the ROWKEY? Might help to have constants for these numbers. Also might help to add the key fields after the start/end. Super nits.

Reply:
This will change in the next PR, so I'm not worried.